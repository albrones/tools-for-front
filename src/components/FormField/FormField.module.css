.root {
  position: relative;
}
.root input {
  background-color: #ffffff;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.42);
  box-shadow: none;
  color: inherit;
  font-size: 14px;
  height: 44px;
  line-height: 44px;
  outline: none;
  padding-left: 10px;
  padding-right: 10px;
  transition: background-color 0.2s ease-in-out;
  width: 100%;
}
.root input:focus,
.root input:hover {
  border-color: rgba(0, 0, 0, 0.87);
}
.root input[type="checkbox"] {
  display: inline-block;
  vertical-align: -16px;
  width: 44px;
}
.root label {
  display: block;
  font-size: 12px;
  font-weight: 400;
}
