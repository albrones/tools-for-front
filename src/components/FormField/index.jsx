import React from "react";
import PropTypes from "prop-types";

import styles from "./FormField.module.css";

class FormField extends React.Component {
  shouldComponentUpdate(nextProps) {
    const compare = i => this.props[i] !== nextProps[i];

    const nextKeys = Object.keys(nextProps);
    const prevKeys = Object.keys(this.props);
    return nextKeys.some(compare) && prevKeys.some(compare);
  }

  render() {
    const { className, label, name, type, defaultValue, onChange, value, ...props } = this.props;
    const inputArgs = { id: `form-field-${name}`, name, type };

    if (value !== undefined) {
      inputArgs.value = value;
    } else {
      inputArgs.defaultValue = defaultValue;
    }

    if (type === "checkbox" || type === "radio") {
      inputArgs.checked = value !== undefined ? value : false;
      inputArgs.onChange = e => onChange(e.target.checked);
      return (
        <div className={`${styles.root} ${className}`}>
          <label htmlFor={`form-field-${name}`}>
            <input {...props} {...inputArgs} />
            {label}
          </label>
        </div>
      );
    }

    inputArgs.onChange = e => onChange(e.target.value);
    return (
      <div className={`${styles.root} ${className}`}>
        <label htmlFor={`form-field-${name}`}>
          {label}
          <input {...inputArgs} />
        </label>
      </div>
    );
  }
}

FormField.defaultProps = {
  defaultValue: "",
  type: "text",
  className: "",
  value: undefined
};

FormField.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  type: PropTypes.string,
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.bool, PropTypes.number]),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool, PropTypes.number])
};

export default FormField;
