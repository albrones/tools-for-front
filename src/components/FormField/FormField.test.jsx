import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import FormField from ".";

Enzyme.configure({ adapter: new Adapter() });

test("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<FormField label="Test" name="test" onChange={() => {}} />, div);
});

test("renders type=text without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<FormField label="Test" name="test" onChange={() => {}} type="text" />, div);
});

test("renders type=number without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<FormField label="Test" name="test" onChange={() => {}} type="number" />, div);
});

test("renders type=checkbox without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<FormField label="Test" name="test" onChange={() => {}} type="checkbox" />, div);
});

test("renders type=radio without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<FormField label="Test" name="test" onChange={() => {}} type="radio" />, div);
});

test("renders className without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <FormField label="Test" name="test" onChange={() => {}} className="test-class" />,
    div
  );
});

test("renders radio + className without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <FormField label="Test" name="test" onChange={() => {}} type="radio" className="test-class" />,
    div
  );
});

test("Update props being applied", () => {
  const component = shallow(
    <FormField label="Test" name="test" onChange={() => {}} type="radio" />
  );

  expect(component.find("label").prop("htmlFor")).toBe("form-field-test");
  expect(component.find("label").text()).toBe("Test");
  expect(component.find("input").prop("type")).toBe("radio");

  component.setProps({
    label: "Test2",
    name: "test2",
    type: "text"
  });

  expect(component.find("label").prop("htmlFor")).toBe("form-field-test2");
  expect(component.find("label").text()).toBe("Test2");
  expect(component.find("input").prop("type")).toBe("text");
});

test("updates values", () => {
  const component = renderer.create(
    <FormField label="Test" name="test" onChange={() => {}} type="text" />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.props.value = "New Value";
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.props.value = "";
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Change checkbox after click", () => {
  const onChange = jest.fn();
  const checkbox = shallow(
    <FormField label="Test" name="test" onChange={onChange} type="checkbox" />
  );

  const input = checkbox.find("input");

  input.simulate("change", {
    target: { checked: input.getElement().props.checked }
  });
  expect(onChange).toHaveBeenCalled();
});

test("Update value", () => {
  const onChange = jest.fn();
  const checkbox = shallow(
    <FormField label="Test" name="test" value={false} onChange={onChange} type="checkbox" />
  );

  expect(checkbox.find("input").prop("checked")).toBe(false);
  checkbox.setProps({ value: true });
  expect(checkbox.find("input").prop("checked")).toBe(true);
});
