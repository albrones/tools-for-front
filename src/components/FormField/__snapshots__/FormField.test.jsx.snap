// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updates values 1`] = `
<div
  className="root "
>
  <label
    htmlFor="form-field-test"
  >
    Test
    <input
      defaultValue=""
      id="form-field-test"
      name="test"
      onChange={[Function]}
      type="text"
    />
  </label>
</div>
`;

exports[`updates values 2`] = `
<div
  className="root "
>
  <label
    htmlFor="form-field-test"
  >
    Test
    <input
      defaultValue=""
      id="form-field-test"
      name="test"
      onChange={[Function]}
      type="text"
    />
  </label>
</div>
`;

exports[`updates values 3`] = `
<div
  className="root "
>
  <label
    htmlFor="form-field-test"
  >
    Test
    <input
      defaultValue=""
      id="form-field-test"
      name="test"
      onChange={[Function]}
      type="text"
    />
  </label>
</div>
`;
